examples/active_replication/flux_producer_operator.py

```diff
@@ -0,0 +1,93 @@
+from erdos.data_stream import DataStream
+from erdos.op import Op
+from erdos.utils import setup_logging
+
+import flux_utils
+from flux_utils import is_ack_stream, is_not_ack_stream, is_not_back_pressure
+from flux_buffer import Buffer
+
+import threading
+
+
+class FluxProducerOperator(Op):
+    def __init__(self,
+                 name,
+                 replica_num,
+                 output_stream_name,
+                 log_file_name=None):
+        super(FluxProducerOperator, self).__init__(name)
+        self._logger = setup_logging(self.name, log_file_name)
+        self._replica_num = replica_num
+        self._ex_secondary = False
+        self._output_stream_name = output_stream_name
+        self._output_seq_num = 0
+        if replica_num > 0:
+            # only replicas have these buffers to keep track
+            # with the progress of primary op
+            self._buffer = Buffer(1)
+            self._ack_buffer = set()
+        self.lock = threading.Lock()
+
+    @staticmethod
+    def setup_streams(input_streams, output_stream_name):
+        input_streams\
+            .filter(is_not_ack_stream)\
+            .filter(is_not_back_pressure)\
+            .add_callback(FluxProducerOperator.on_msg)
+        input_streams.filter(is_ack_stream).add_callback(
+            FluxProducerOperator.on_ack_msg)
+
+        return [DataStream(name=output_stream_name)]
+
+    def on_msg(self, msg):
+        self.lock.acquire()
+        # XXX(ionel): Need to ensure that we remove output_seq_num
+        # during recovery.
+        if self._replica_num == 0:  # Primary
+            msg.data = (self._output_seq_num, msg.data)
+            if self._ex_secondary:  # used to a secondary
+                if len(self._ack_buffer) > 0 and max(self._ack_buffer) > self._output_seq_num:   # behind progress
+                    lst = [str(a) for a in self._ack_buffer]
+                    lst = ",".join(lst)
+                    assert self._output_seq_num in self._ack_buffer, str(self._output_seq_num) + " not in " + lst
+                    self._ack_buffer.remove(self._output_seq_num)
```

Comment:
Looking at this code, I think it's much easier if you transform ack_buffer into a queue. ACK_buffer will store the ACKS in increasing order, and we will also always remove the first ACK because we'll also  receive messages in order from the upstream operator.